app/src/main/java/com/amaze/filemanager/ui/fragments/FtpServerFragment.kt

```diff
@@ -364,12 +367,77 @@ class FtpServerFragment : Fragment(R.layout.fragment_ftp) {
         updateStatus()
     }
 
+    private fun createOpenDocumentTreeIntentCallback(callback: (directoryUri: Uri) -> Unit):
+        ActivityResultLauncher<Intent> {
+            return registerForActivityResult(
+                ActivityResultContracts.StartActivityForResult()
+            ) {
+                if (it.resultCode == RESULT_OK && Build.VERSION.SDK_INT >= LOLLIPOP) {
+                    val directoryUri = it.data?.data ?: return@registerForActivityResult
+                    requireContext().contentResolver.takePersistableUriPermission(
+                        directoryUri,
+                        Intent.FLAG_GRANT_READ_URI_PERMISSION
+                            or Intent.FLAG_GRANT_WRITE_URI_PERMISSION
+                    )
+                    callback.invoke(directoryUri)
+                }
+            }
+        }
+
+    /** Check URI access if  */
+    private fun checkUriAccessIfNecessary(callback: () -> Unit) {
+        if (Build.VERSION.SDK_INT >= M) {
+            val directoryUri = mainActivity.prefs.getString(KEY_PREFERENCE_PATH, DEFAULT_PATH)
+            Uri.parse(directoryUri).run {
+                if (requireContext().checkUriPermission(
+                        this, Process.myPid(), Process.myUid(),
+                        Intent.FLAG_GRANT_WRITE_URI_PERMISSION
+                            or Intent.FLAG_GRANT_READ_URI_PERMISSION
+                    ) == PackageManager.PERMISSION_DENIED
+                ) {
+                    mainActivity.accent.run {
```

Comment:
Please at least put this function in a named constants so that this is more readable.